### PR TITLE
Feature/tzip 16

### LINKED ISSUE #999: Handling tzip-16 compatible contracts
As discussed with @bhaskarkishore metadata regarding tzip-16 contracts should be stored in the `known_addresses` table which now will look like this:
```
CREATE TABLE tezos.known_addresses (
    address character varying NOT NULL,
    alias character varying,
    name character varying,
    description character varying,
    version character varying,
    license_name character varying,
    license_details character varying,
    authors character varying,
    homepage character varying,
    source_tools character varying,
    source_location character varying,
    interfaces character varying,
    block_level bigint,
    timestamp timestamp without time zone,
    metadata_uri character varying,
    views character varying
);
```

The basic algorithm is rather simple to explain:
1. Parse operations parameters which match the tzip16 schema
2. Retrieve the metadata from URL (http/ipfs for now)
3. Store metadata in the modified `known_addresses` table.

@bhaskarkishore Is it ok, or am I missing something?

### PR DESCRIPTION
**Pull request by piotrkosecki**
Resolves #999 

Implemented handling of the tzip-16 metadata fetching as described in the ticket.

### CODE REVIEW

File: sql/conseil.sql [hunk 1]
```diff
@@ -128,7 +128,34 @@ CREATE TABLE tezos.registered_tokens (
...
+    metadata_big_map_id integer NOT NULL,
+    metadata_big_map_type text NOT NULL,
+    metadata_path text NOT NULL
+);
+
+CREATE TABLE tezos.metadata (
+    address text NOT NULL,
+    raw_metadata text NOT NULL,
+    name text NOT NULL,
+    desctiption text,
```
Comment (vishakh) [inline sql/conseil.sql:145]: Typo

**Review by vishakh — CHANGES_REQUESTED** [review]

File: sql/conseil.sql (hunk 1 above)
Comment (piotrkosecki) [reply sql/conseil.sql]: fixed

File: conseil-lorre/src/main/scala/tech/cryptonomic/conseil/indexer/tezos/processing/MetadataProcessor.scala [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+import cats.syntax.applicative._
+import cats.implicits._
+import tech.cryptonomic.conseil.common.tezos
+import slick.jdbc.PostgresProfile.api._
+import tech.cryptonomic.conseil.common.io.Logging.ConseilLogSupport
+
+import scala.concurrent.{ExecutionContext, Future}
+import scala.util.Try
+
+class MetadataProcessor(
```
Comment (vishakh) [inline conseil-lorre/src/main/scala/tech/cryptonomic/conseil/indexer/tezos/processing/MetadataProcessor.scala:27]: A comment explaining the purpose of this class and how it works would be very useful here.

File: conseil-lorre/src/main/scala/tech/cryptonomic/conseil/indexer/tezos/michelson/contracts/TokenContracts.scala [hunk 3]
```diff
@@ -364,4 +369,65 @@ object TokenContracts extends ConseilLogSupport {
     }
 
   }
+
+  object Tzip16 {
+
+    private def proceduralDecode(hex: String): String = {
```
Comment (vishakh) [inline conseil-lorre/src/main/scala/tech/cryptonomic/conseil/indexer/tezos/michelson/contracts/TokenContracts.scala:375]: Please leave a note about the function here.

File: conseil-lorre/src/main/scala/tech/cryptonomic/conseil/indexer/tezos/Tzip16MetadataFetcher.scala [hunk 4]
```diff
@@ -0,0 +1,468 @@
...
+    fetch[(InternalOperationResults.Transaction, String), Option[(String, Tzip16Metadata)], Future, List, Throwable]
+      .run(addresses)
+
+  def getMetadataWithRegisteredTokensRow(
+      addresses: List[(Tables.RegisteredTokensRow, Tables.BigMapContentsRow, String)]
+  ): Future[List[((Tables.RegisteredTokensRow, Tables.BigMapContentsRow, String), Option[(String, Tzip16Metadata)])]] =
+    fetch[(Tables.RegisteredTokensRow, Tables.BigMapContentsRow, String), Option[(String, Tzip16Metadata)], Future, List, Throwable]
+      .run(addresses)
+
+  def getHttpMetadataWithRegisteredTokensRow(
```
Comment (vishakh) [inline conseil-lorre/src/main/scala/tech/cryptonomic/conseil/indexer/tezos/Tzip16MetadataFetcher.scala:92]: Please remove.

File: conseil-lorre/src/main/scala/tech/cryptonomic/conseil/indexer/tezos/Tzip16MetadataFetcher.scala [hunk 5]
```diff
@@ -0,0 +1,468 @@
...
+      Kleisli(
+        fetchKeys => {
+          val hashes = fetchKeys
+          logger.info("Fetching tzip metadata")
+          node.runBatchedGetQuery(fetchKeys, makeUrl, 2).onError {
+            case err =>
+              val showHashes = hashes.mkString(", ")
+              logger
+                .error(
+                  s"I encountered problems while fetching baking rights from TZIP-16 for blocks $showHashes. The error says ${err.getMessage}"
```
Comment (vishakh) [inline conseil-lorre/src/main/scala/tech/cryptonomic/conseil/indexer/tezos/Tzip16MetadataFetcher.scala:136]: Fix please

File: conseil-lorre/src/main/scala/tech/cryptonomic/conseil/indexer/tezos/processing/MetadataProcessor.scala [hunk 6]
```diff
@@ -0,0 +1,167 @@
...
+          val nftAddresses = nfts.foldLeft(
+            (
+              List.empty[(Tables.RegisteredTokensRow, Tables.BigMapContentsRow, String)],
+              List.empty[(Tables.RegisteredTokensRow, Tables.BigMapContentsRow, String)]
+            )
+          ) {
+            case ((ipfs, http), acc) if acc._3.startsWith("ipfs") => (acc :: ipfs, http)
+            case ((ipfs, http), acc) if acc._3.startsWith("http") => (ipfs, acc :: http)
+          }
+          println(metadata)
```
Comment (vishakh) [inline conseil-lorre/src/main/scala/tech/cryptonomic/conseil/indexer/tezos/processing/MetadataProcessor.scala:116]: Remove please.

**Review by vishakh — CHANGES_REQUESTED** [review]

**Review by vishakh — APPROVED** [review]

**Pull request merged**